### PR TITLE
Fix for build under Windows with MSYS2 - gcc version 4.9.2 (GCC) Target: x86_64-pc-msys + ure.sub

### PR DESCRIPTION
**Pull request by dmazzella**
- Fix for build under Windows with MSYS2

$ uname
MSYS_NT-10.0

$ gcc --version
gcc (GCC) 4.9.2
Copyright (C) 2014 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
- Added sub into modure and relative tests

### CODE REVIEW

File: extmod/modmachine.c [hunk 1]
```diff
@@ -37,7 +37,9 @@
 
 #if MICROPY_PY_MACHINE
 
+#ifndef PAGE_SIZE
 #define PAGE_SIZE 4096
+#endif
```
Comment (dpgeorge) [inline extmod/modmachine.c]: I don't know if this is the correct way to do it.  PAGE_SIZE may already be defined, but it may not be defined to the correct thing.

I think the right way to do it is use getpagesize(), which is supposed to give you the page size used by mmap.

File: unix/main.c [hunk 2]
```diff
@@ -416,7 +416,7 @@ int main(int argc, char **argv) {
             } else if (strcmp(argv[a], "-v") == 0) {
                 mp_verbose_flag++;
             } else if (strncmp(argv[a], "-O", 2) == 0) {
-                if (isdigit(argv[a][2])) {
+                if (isdigit((byte)argv[a][2])) {
```
Comment (dpgeorge) [inline unix/main.c]: As we already discussed, this is something that needs to be fixed by the provider of isdigit.

Alternatively, we can use our own: unichar_isdigit.

**Comment by dpgeorge** [comment]
The implementation of re_sub seems quite long-winded and it'll take some time to review exactly how it works.

As for the re.sub tests, they need to be simpler.  The tests need to be approx 1 line long.  There needs to be lots of them, each one testing a specific feature or corner case.  Take a look at other tests to see how it's done.

File: extmod/modmachine.c (hunk 1 above)
Comment (pfalcon) [inline extmod/modmachine.c]: Reading man for getpagesize, there's many steps how it can be done even "righter", and that's exactly why I did it KISS way like that. I just pushed fix which should avoid naming conflicts. But note that if MSYS2 doesn't provide mmap(), the modmachine shouldn't be enabled for it (won't work as expected anyway).

**Comment by pfalcon** [comment]
@dmazzella : This is better, but there's still problem that despite title given, it contains unrelated changes. No big problem, but some issues would be fixed differently than you propose, so you would need to use rebase interactive and cherry-pick to update your code, and that's what required to provide separate PR for different issues in the first place.

File: extmod/modmachine.c (hunk 1 above)
Comment (dpgeorge) [inline extmod/modmachine.c]: Agreed, simpler is better for this case.

**Comment by pfalcon** [comment]
But the main issue - what's usecase behind this implementation of .sub() ? There's quite a lot of not-so-easy to understand code (which has lot of codestyle issues, but that's minor comparing to the issue discussed). Who would need that code and for what? Does it justify pushing that code on everyone else? Do you yourself need all those features - \g<> groups, callable support? The way uPy is implemented is that we start with minimal code, and add more features one by one on as-needed basis. And doing it on right level is pretty important. "ure" module provides the required minimum of functions dealing with regex'es. Anything else can be implemented on top of that in Python, and that was already done for PCRE implementation and can be reused for "ure": https://github.com/micropython/micropython-lib/blob/master/re-pcre/re.py#L83

**Comment by dmazzella** [comment]
@pfalcon "Do you yourself need all those features - \g<> groups, callable support?" 
for my project i need callable support,  \g<> groups isn't a need but a plus to have

And I know .sub can be written in python, I do not know, for me it is good to have in .sub "ure" but the decision is yours.

**Comment by dmazzella** [comment]
@dpgeorge I have simplified tests to have a minimum set of functionality

**Comment by pfalcon** [comment]
> And I know .sub can be written in python, for me is good to have .sub in "ure" but the decision is yours.

Yes, I agree it would be good to have C implementation, but it should be maintainable and there should be clear understanding how many bytes is worth to spend on it. As @dpgeorge points out, it will take a while to understand how that code works (and prove that it works correctly!), and I don't think .sub() is so frequently used function that it's worth to spend more than 200 bytes on it, and this code clearly takes more. (But bytecode Python function will take around that, and can be used only by folks who actually need it.)
